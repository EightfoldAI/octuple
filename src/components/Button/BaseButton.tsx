import React, { FC, Ref } from 'react';
import {
    ButtonIconAlign,
    ButtonShape,
    ButtonSize,
    ButtonTextAlign,
    ButtonTheme,
    ButtonWidth,
    InternalButtonProps,
    SplitButton,
} from './';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';

import styles from './button.module.scss';

export const BaseButton: FC<InternalButtonProps> = React.forwardRef(
    (
        {
            alignIcon = ButtonIconAlign.Left,
            alignText = ButtonTextAlign.Center,
            allowDisabledFocus = false,
            ariaLabel,
            buttonWidth = ButtonWidth.fitContent,
            checked = false,
            classNames,
            counter,
            disabled = false,
            disruptive = false,
            dropShadow = false,
            htmlType,
            iconProps,
            id,
            onClick,
            onContextMenu,
            shape = ButtonShape.Pill,
            size = ButtonSize.Flex,
            split,
            splitButtonChecked = false,
            splitButtonProps,
            style,
            text,
            theme,
            toggle,
            type,
            ...rest
        },
        ref: Ref<HTMLButtonElement>
    ) => {
        const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
        const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
        const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
        const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

        const counterExists: boolean = !!counter;
        const iconExists: boolean = !!iconProps;
        const textExists: boolean = !!text;

        const buttonBaseSharedClassNames: string = mergeClasses([
            classNames,
            {
                [styles.buttonSize3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonSize2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonSize1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonSize1]: size === ButtonSize.Large },
            { [styles.buttonSize2]: size === ButtonSize.Medium },
            { [styles.buttonSize3]: size === ButtonSize.Small },
            { [styles.pillShape]: shape === ButtonShape.Pill },
            { [styles.dropShadow]: dropShadow },
            { [styles.dark]: theme === ButtonTheme.dark },
        ]);

        const buttonBaseClassNames: string = mergeClasses([
            buttonBaseSharedClassNames,
            { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
            { [styles.splitLeft]: split },
            { [styles.left]: alignText === ButtonTextAlign.Left },
            { [styles.right]: alignText === ButtonTextAlign.Right },
            {
                [styles.iconLeft]:
                    iconExists && alignIcon === ButtonIconAlign.Left,
            },
            {
                [styles.iconRight]:
                    iconExists && alignIcon === ButtonIconAlign.Right,
            },
            { [styles.disabled]: allowDisabledFocus || disabled },
        ]);

        const buttonTextClassNames: string = mergeClasses([
            {
                [styles.buttonText3]:
                    size === ButtonSize.Flex && largeScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && mediumScreenActive,
            },
            {
                [styles.buttonText2]:
                    size === ButtonSize.Flex && smallScreenActive,
            },
            {
                [styles.buttonText1]:
                    size === ButtonSize.Flex && xSmallScreenActive,
            },
            { [styles.buttonText1]: size === ButtonSize.Large },
            { [styles.buttonText2]: size === ButtonSize.Medium },
            { [styles.buttonText3]: size === ButtonSize.Small },
        ]);

        const badgeClassNames: string = mergeClasses([
            styles.counter,
            buttonTextClassNames,
        ]);

        const getButtonIconSize = (): IconSize => {
            let iconSize: IconSize;
            if (size === ButtonSize.Flex && largeScreenActive) {
                iconSize = IconSize.Small;
            } else if (
                size === ButtonSize.Flex &&
                (mediumScreenActive || smallScreenActive)
            ) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Flex && xSmallScreenActive) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Large) {
                iconSize = IconSize.Large;
            } else if (size === ButtonSize.Medium) {
                iconSize = IconSize.Medium;
            } else if (size === ButtonSize.Small) {
                iconSize = IconSize.Small;
            }
            return iconSize;
        };

        const getButtonIcon = (): JSX.Element => (
            <Icon
                {...iconProps}
                classNames={mergeClasses([styles.icon, iconProps.classNames])}
                size={getButtonIconSize()}
            />
        );

        const getButtonContent = (
            buttonTextClassNames: string,
            text: string
        ): JSX.Element => (
            <span className={buttonTextClassNames}>
                {text ? text : 'Button'}
                {counterExists && (
                    <Badge classNames={badgeClassNames}>{counter}</Badge>
                )}
            </span>
        );

        return (
            <>
                <button
                    ref={ref}
                    {...rest}
                    aria-checked={toggle ? !!checked : undefined}
                    aria-disabled={allowDisabledFocus}
                    aria-label={ariaLabel}
                    aria-pressed={toggle ? !!checked : undefined}
                    defaultChecked={checked}
                    disabled={disabled}
                    className={buttonBaseClassNames}
                    id={id}
                    onClick={!allowDisabledFocus ? onClick : null}
                    style={style}
                    type={htmlType}
                >
                    {iconExists && !textExists && getButtonIcon()}
                    {counterExists && !textExists && (
                        <Badge classNames={badgeClassNames}>{counter}</Badge>
                    )}
                    {iconExists && textExists && (
                        <span>
                            {getButtonIcon()}
                            {getButtonContent(buttonTextClassNames, text)}
                        </span>
                    )}
                    {!iconExists &&
                        getButtonContent(buttonTextClassNames, text)}
                </button>
                {split && (
                    <SplitButton
                        {...splitButtonProps}
                        classNames={
                            buttonBaseSharedClassNames +
                            ' ' +
                            splitButtonProps?.classNames
                        }
                        checked={splitButtonChecked}
                        disruptive={disruptive}
                        dropShadow={dropShadow}
                        onClick={
                            !splitButtonProps?.allowDisabledFocus
                                ? onContextMenu
                                : null
                        }
                        shape={shape}
                        size={size}
                        split={split}
                        theme={theme}
                        type={type}
                    />
                )}
            </>
        );
    }
);
