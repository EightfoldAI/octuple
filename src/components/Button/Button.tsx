import React, { FC, Ref, useContext, useEffect, useRef, useState } from 'react';
import DisabledContext, { Disabled } from '../ConfigProvider/DisabledContext';
import { SizeContext, Size } from '../ConfigProvider';
import {
  ButtonIconAlign,
  ButtonProps,
  ButtonShape,
  ButtonSize,
  ButtonTextAlign,
  ButtonVariant,
  ButtonWidth,
  SplitButton,
} from '.';
import { Icon, IconSize } from '../Icon';
import { Badge } from '../Badge';
import { InnerNudge, NudgeAnimation, NudgeProps } from './Nudge';
import { Breakpoints, useMatchMedia } from '../../hooks/useMatchMedia';
import { mergeClasses } from '../../shared/utilities';
import { Loader, LoaderSize } from '../Loader';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { useMergedRefs } from '../../hooks/useMergedRefs';
import { useNudge } from './Nudge/Hooks/useNudge';

import styles from './button.module.scss';

export const Button: FC<ButtonProps> = React.forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      alignIcon = ButtonIconAlign.Left,
      alignText = ButtonTextAlign.Center,
      allowDisabledFocus = false,
      ariaLabel,
      buttonWidth = ButtonWidth.fitContent,
      checked = false,
      classNames,
      configContextProps = {
        noDisabledContext: false,
        noSizeContext: false,
      },
      counter,
      disabled = false,
      disruptive = false,
      dropShadow = false,
      floatingButtonProps,
      nudgeProps: defaultNudgeProps,
      htmlType,
      iconProps,
      id,
      loading = false,
      onClick,
      onContextMenu,
      prefixIconProps,
      shape = ButtonShape.Pill,
      size = ButtonSize.Medium,
      split,
      splitButtonChecked = false,
      splitButtonProps,
      style,
      text,
      toggle,
      transparent = false,
      type, // TODO: Remove in Octuple v3.0.0 and use `variant` only.
      variant = ButtonVariant.Default,
      ...rest
    } = props;
    const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);
    const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
    const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
    const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);

    const internalRef: React.MutableRefObject<HTMLButtonElement> =
      useRef<HTMLButtonElement>(null);

    const mergedRef = useMergedRefs(internalRef, ref);

    const htmlDir: string = useCanvasDirection();

    const contextuallyDisabled: Disabled = useContext(DisabledContext);
    const mergedDisabled: boolean = configContextProps.noDisabledContext
      ? disabled
      : contextuallyDisabled || disabled;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = configContextProps.noSizeContext
      ? size
      : contextuallySized || size;

    const counterExists: boolean = !!counter;
    const iconExists: boolean = !!iconProps;
    const prefixIconExists: boolean = !!prefixIconProps;
    const textExists: boolean = !!text;

    const [nudgeProps, setNudgeProps] = useState<NudgeProps>(defaultNudgeProps);
    const innerNudgeRef: React.MutableRefObject<HTMLSpanElement> =
      useRef<HTMLSpanElement>(null);

    useEffect(() => {
      setNudgeProps(
        props.nudgeProps
          ? props.nudgeProps
          : {
              animation: NudgeAnimation.Background,
              delay: 2000,
              iterations: 1,
              enabled: false,
            }
      );
    }, [nudgeProps?.enabled]);

    useNudge(disruptive, nudgeProps, [internalRef, innerNudgeRef], styles);

    // TODO: Remove in Octuple v3.0.0 and use `variant` only.
    // For now, if `type` has a value use it, else use `variant`.
    const mergedVariant: ButtonVariant = !!type ? type : variant;

    const buttonSharedClassNames: string = mergeClasses([
      classNames,
      styles.button,
      { [styles.buttonDefault]: mergedVariant === ButtonVariant.Default },
      { [styles.buttonNeutral]: mergedVariant === ButtonVariant.Neutral },
      { [styles.buttonPrimary]: mergedVariant === ButtonVariant.Primary },
      { [styles.buttonSecondary]: mergedVariant === ButtonVariant.Secondary },
      { [styles.buttonSystemUi]: mergedVariant === ButtonVariant.SystemUI },
      {
        [styles.buttonDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Default,
      },
      {
        [styles.buttonPrimaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Primary,
      },
      {
        [styles.buttonSecondaryDisruptive]:
          disruptive && mergedVariant === ButtonVariant.Secondary,
      },
      {
        [styles.transparent]:
          transparent && mergedVariant === ButtonVariant.SystemUI,
      },
      {
        [styles.buttonSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonSmall]: mergedSize === ButtonSize.Small },
      { [styles.pillShape]: shape === ButtonShape.Pill },
      {
        [styles.roundShape]:
          shape === ButtonShape.Round && !split && !textExists,
      },
      { [styles.dropShadow]: dropShadow },
      { [styles.disabled]: allowDisabledFocus || mergedDisabled },
      { [styles.floating]: floatingButtonProps?.enabled },
      { [styles.buttonRtl]: htmlDir === 'rtl' },
      {
        [styles.buttonConic]:
          !disruptive &&
          nudgeProps?.enabled &&
          nudgeProps?.animation === NudgeAnimation.Conic,
      },
    ]);

    const buttonClassNames: string = mergeClasses([
      buttonSharedClassNames,
      { [styles.buttonStretch]: buttonWidth === ButtonWidth.fill },
      { [styles.splitLeft]: split },
      { [styles.left]: alignText === ButtonTextAlign.Left },
      { [styles.right]: alignText === ButtonTextAlign.Right },
      {
        [styles.iconLeft]: iconExists && alignIcon === ButtonIconAlign.Left,
      },
      {
        [styles.iconRight]: iconExists && alignIcon === ButtonIconAlign.Right,
      },
      { [styles.loading]: loading },
    ]);

    const buttonTextClassNames: string = mergeClasses([
      {
        [styles.buttonTextSmall]:
          mergedSize === ButtonSize.Flex && largeScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && mediumScreenActive,
      },
      {
        [styles.buttonTextMedium]:
          mergedSize === ButtonSize.Flex && smallScreenActive,
      },
      {
        [styles.buttonTextLarge]:
          mergedSize === ButtonSize.Flex && xSmallScreenActive,
      },
      { [styles.buttonTextLarge]: mergedSize === ButtonSize.Large },
      { [styles.buttonTextMedium]: mergedSize === ButtonSize.Medium },
      { [styles.buttonTextSmall]: mergedSize === ButtonSize.Small },
    ]);

    const badgeClassNames: string = mergeClasses([
      styles.counter,
      buttonTextClassNames,
    ]);

    const getButtonIconSize = (): IconSize => {
      let iconSize: IconSize;
      if (mergedSize === ButtonSize.Flex && largeScreenActive) {
        iconSize = IconSize.Small;
      } else if (
        mergedSize === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Flex && xSmallScreenActive) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Large) {
        iconSize = IconSize.Large;
      } else if (mergedSize === ButtonSize.Medium) {
        iconSize = IconSize.Medium;
      } else if (mergedSize === ButtonSize.Small) {
        iconSize = IconSize.Small;
      }
      return iconSize;
    };

    const getLoaderSize = (): LoaderSize => {
      let loaderSize: LoaderSize;
      if (size === ButtonSize.Flex && largeScreenActive) {
        loaderSize = LoaderSize.Small;
      } else if (
        size === ButtonSize.Flex &&
        (mediumScreenActive || smallScreenActive)
      ) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Flex && xSmallScreenActive) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Large) {
        loaderSize = LoaderSize.Large;
      } else if (size === ButtonSize.Medium) {
        loaderSize = LoaderSize.Medium;
      } else if (size === ButtonSize.Small) {
        loaderSize = LoaderSize.Small;
      }
      return loaderSize;
    };

    const getButtonLoader = (): JSX.Element =>
      loading && (
        <Loader
          classNames={styles.loader}
          dotClassNames={styles.loaderDot}
          size={getLoaderSize()}
        />
      );

    const getButtonIcon = (): JSX.Element => (
      <Icon
        {...iconProps}
        classNames={mergeClasses([styles.icon, iconProps.classNames])}
        size={getButtonIconSize()}
      />
    );

    const getPrefixIcon = (): JSX.Element => (
      <Icon
        {...prefixIconProps}
        classNames={mergeClasses([
          styles.icon,
          styles.prefixIcon,
          prefixIconProps.classNames,
        ])}
        size={getButtonIconSize()}
      />
    );

    const getButtonContent = (
      buttonTextClassNames: string,
      text: string
    ): JSX.Element => (
      <span className={buttonTextClassNames}>
        {text ? text : 'Button'}
        {counterExists && <Badge classNames={badgeClassNames}>{counter}</Badge>}
      </span>
    );

    return (
      <>
        <button
          {...rest}
          ref={mergedRef}
          aria-checked={toggle ? !!checked : undefined}
          aria-disabled={mergedDisabled || loading}
          aria-label={ariaLabel}
          aria-pressed={toggle ? !!checked : undefined}
          defaultChecked={checked}
          disabled={(!allowDisabledFocus && mergedDisabled) || loading}
          className={buttonClassNames}
          id={id}
          onClick={!allowDisabledFocus ? onClick : null}
          style={style}
          type={htmlType}
        >
          <InnerNudge
            classNames={mergeClasses([
              styles.innerNudge,
              {
                [styles.conic]: nudgeProps?.animation === NudgeAnimation.Conic,
              },
            ])}
            disruptive={disruptive}
            id={id ? `${id}-nudge` : 'button-nudge'}
            nudgeProps={nudgeProps}
            ref={innerNudgeRef}
            style={style}
          />
          {iconExists && prefixIconExists && !textExists && (
            <span>
              {getButtonIcon()}
              {getPrefixIcon()}
            </span>
          )}
          {iconExists &&
            !textExists &&
            !prefixIconExists &&
            !counterExists &&
            getButtonIcon()}
          {counterExists && !textExists && !loading && !iconExists && (
            <Badge classNames={badgeClassNames}>{counter}</Badge>
          )}
          {iconExists && counterExists && !textExists && !loading && (
            <span>
              {getButtonIcon()}
              <Badge classNames={badgeClassNames}>{counter}</Badge>
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {iconExists && textExists && (
            <span>
              {getButtonIcon()}
              {getButtonContent(buttonTextClassNames, text)}
              {prefixIconExists && getPrefixIcon()}
            </span>
          )}
          {!iconExists && getButtonContent(buttonTextClassNames, text)}
          {getButtonLoader()}
        </button>
        {split && (
          <SplitButton
            {...splitButtonProps}
            classNames={
              buttonSharedClassNames + ' ' + splitButtonProps?.classNames
            }
            checked={splitButtonChecked}
            disruptive={disruptive}
            dropShadow={dropShadow}
            nudgeProps={nudgeProps}
            onClick={
              !splitButtonProps?.allowDisabledFocus ? onContextMenu : null
            }
            shape={shape}
            size={mergedSize}
            split={split}
            variant={mergedVariant}
          />
        )}
      </>
    );
  }
);
