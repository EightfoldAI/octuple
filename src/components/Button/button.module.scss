.button {
    --button-default-foreground-color: var(--primary-color);
    --button-focus-outline-color: var(--blueviolet-color-50);
    --button-primary-default-border-color: var(--primary-color);
    --button-two-state-default-foreground-color: var(--grey-color-70);
    --button-two-state-hover-foreground-color: var(--primary-color);
    --button-two-state-focus-foreground-color: var(--primary-color-80);
    --button-two-state-checked-foreground-color: var(--primary-color-80);
    --button-two-state-active-foreground-color: var(--primary-color-80);
    --button-two-state-visited-foreground-color: var(--primary-color-80);
    --button-two-state-default-background-color: var(--white-color);
    --button-two-state-hover-background-color: var(--white-color);
    --button-two-state-focus-background-color: var(--white-color);
    --button-two-state-checked-background-color: var(--primary-color-20);
    --button-two-state-active-background-color: var(--primary-color-20);
    --button-two-state-visited-background-color: var(--primary-color-20);
    --button-counter-default-background-color: var(--grey-color-20);
    --button-counter-hover-background-color: var(--primary-color-20);
    --button-counter-checked-background-color: var(--primary-color-10);
    --button-counter-focus-background-color: var(--primary-color-20);
    --button-counter-active-background-color: var(--primary-color-10);
    --button-counter-visited-background-color: var(--primary-color-10);
    --button-two-state-default-outline-color: var(--grey-color-70);
    --button-two-state-hover-outline-color: var(--primary-color-60);
    --button-two-state-checked-outline-color: var(--primary-color);
    --button-two-state-focus-outline-color: var(--primary-color-40);
    --button-two-state-active-outline-color: var(--primary-color);
    --button-two-state-visited-outline-color: var(--primary-color);
    background-color: inherit;
    border-radius: $corner-radius-s;
    cursor: pointer;
    display: inline-block;
    min-height: max-content;
    min-width: max-content;
    transition: all $motion-duration-extra-fast $motion-easing-easeinout 0s;
    white-space: nowrap;

    span:not(.column):not(.counter) {
        align-items: center;
        display: flex;
        flex-direction: row;
    }

    .icon + .button-text-1:not(:empty) {
        margin-left: $button-spacer-large;
    }

    .icon + .button-text-2:not(:empty) {
        margin-left: $button-spacer-medium;
    }

    .icon + .button-text-3:not(:empty) {
        margin-left: $button-spacer-small;
    }

    &.pill-shape {
        border-radius: $corner-radius-xl;
    }

    &.drop-shadow {
        box-shadow: 0 1px 2px rgba(15, 20, 31, 0.12),
            0 2px 8px rgba(15, 20, 31, 0.16);
    }

    &.split-left {
        border-bottom-right-radius: 0;
        border-top-right-radius: 0;

        &.button-stretch {
            &.button-size-1 {
                width: calc(100% - 52px);
            }

            &.button-size-2 {
                width: calc(100% - 44px);
            }

            &.button-size-3 {
                width: calc(100% - 36px);
            }
        }
    }

    &.split-right {
        border-bottom-left-radius: 0;
        border-top-left-radius: 0;
        vertical-align: top;
    }

    &.button-stretch {
        width: 100%;

        span {
            justify-content: center;
        }
    }

    &.left {
        span {
            justify-content: left;
        }
    }

    &.right {
        span {
            justify-content: right;
        }
    }

    &.icon-left {
        span {
            transform: none;
        }
    }

    &.icon-right {
        span:not(.counter) {
            transform: scaleX(-1);
        }

        &.left {
            span {
                justify-content: right;
            }
        }

        &.right {
            span {
                justify-content: left;
            }
        }
    }

    .counter {
        background-color: var(--button-counter-default-background-color);
        display: inline-block;
        padding: 0 $space-xxs;
        text-align: center;
        width: fit-content;
    }

    &:disabled,
    &.disabled {
        opacity: $disabled-alpha-value;
        cursor: not-allowed;
    }

    &:hover {
        .counter {
            background-color: var(--button-counter-hover-background-color);
        }
    }

    &:active {
        .counter {
            background-color: var(--button-counter-active-background-color);
        }
    }

    &:focus {
        outline: 2px solid var(--button-focus-outline-color);
        outline-offset: -2px;

        .counter {
            background-color: var(--button-counter-focus-background-color);
        }
    }
}

.button-text-1 {
    // Button 1 (Mobile L)
    font-family: 'Source Sans Pro', sans-serif;
    font-weight: $text-font-weight-semibold;
    font-size: $text-font-size-5;
    line-height: $text-line-height-3;

    &.counter {
        font-size: $text-font-size-4;
        height: 24px;
        line-height: 24px;
        margin-left: $space-xs;
    }

    &.open-sans {
        font-family: 'Open Sans', sans-serif;
    }
}

.button-text-2 {
    // Button 2 (Mobile M, Desktop L)
    font-family: 'Source Sans Pro', sans-serif;
    font-weight: $text-font-weight-semibold;
    font-size: $text-font-size-3;
    line-height: $text-line-height-2;

    &.counter {
        font-size: $text-font-size-2;
        height: 18px;
        line-height: 18px;
        margin-left: $space-xs;
    }

    &.open-sans {
        font-family: 'Open Sans', sans-serif;
    }
}

.button-text-3 {
    // Button 3 (Desktop M)
    font-family: 'Source Sans Pro', sans-serif;
    font-weight: $text-font-weight-semibold;
    font-size: $text-font-size-2;
    line-height: $text-line-height-1;

    &.counter {
        font-size: $text-font-size-1;
        height: 16px;
        line-height: 16px;
        margin-left: $space-xxs;
    }

    &.open-sans {
        font-family: 'Open Sans', sans-serif;
    }
}

.button-size-1 {
    height: 44px;
    padding: $button-padding-vertical-large $button-padding-horizontal-large;
}

.button-size-2 {
    height: 36px;
    padding: $button-padding-vertical-medium $button-padding-horizontal-medium;
}

.button-size-3 {
    height: 28px;
    padding: $button-padding-vertical-small $button-padding-horizontal-small;
}

.button-primary {
    --bg: var(--accent-color-20);
    --color: var(--primary-color-80);
    color: var(--color);
    background-color: var(--bg);

    &:hover {
        --bg: var(--accent-color-10);
    }

    &:active {
        --bg: var(--accent-color-30);
    }
}

.button-secondary {
    --outline: var(--accent-color-60);
    --bg: transparent;
    --color: var(--primary-color);
    color: var(--color);
    background-color: var(--bg);
    outline: var(--outline) solid 1px;
    outline-offset: -1px;

    &:hover {
        --bg: var(--accent-color-10);
    }

    &:active {
        --bg: var(--accent-color-20);
    }
}

.button-primary-disruptive {
    --bg: var(--disruptive-color-20);
    --color: var(--disruptive-color-80);
    color: var(--color);
    background-color: var(--bg);

    &:hover {
        --bg: var(--disruptive-color-10);
    }

    &:active {
        --bg: var(--disruptive-color-30);
    }
}

.button-secondary-disruptive {
    --bg: transparent;
    --color: var(--disruptive-color-70);
    color: var(--color);
    background-color: var(--bg);
    outline: var(--disruptive-color-70) solid 1px;
    outline-offset: -1px;

    &:hover {
        --bg: var(--disruptive-color-10);
    }

    &:active {
        --bg: var(--disruptive-color-20);
    }
}

.button-default {
    --outline: var(--accent-color-60);
    --bg: var(--accent-color-10);
    --color: var(--primary-color);
    color: var(--color);
    background-color: var(--bg);
    outline: transparent solid 2px;
    outline-offset: -2px;

    &:hover {
        --bg: transparent;
    }

    &:active {
        --bg: var(--accent-color-20);
    }
}

.button-disruptive {
    background-color: var(--white-color);
    color: var(--disruptive-color-70);
    outline: transparent solid 2px;
    outline-offset: -2px;

    &:hover {
        background-color: var(--disruptive-color-20);
        color: var(--disruptive-color-70);
        outline-color: var(--disruptive-color-20);
    }

    &:active {
        background-color: var(--disruptive-color-10);
        color: var(--disruptive-color-80);
        outline-color: var(--disruptive-color-10);
    }

    &:focus {
        background-color: var(--disruptive-color-10);
        color: var(--disruptive-color-80);
        outline-color: var(--disruptive-color-80);
    }
}

.button-neutral {
    --bg: var(--grey-color-10);
    --color: var(--grey-color-70);
    color: var(--color);
    background-color: var(--bg);
    outline: transparent solid 2px;
    outline-offset: -2px;

    &:hover {
        --bg: transparent;
    }

    &:active {
        --bg: var(--grey-color-20);
    }
}

.split-button {
    position: relative;

    .split-divider {
        position: absolute;
        width: 1px;

        &.split-divider-primary {
            background-color: var(--text-inverse-color);
        }

        &.split-divider-secondary {
            background-color: var(--button-default-foreground-color);
        }

        &.split-divider-primary-disruptive {
            background-color: var(--text-inverse-color);
        }

        &.split-divider-secondary-disruptive {
            background-color: var(--disruptive-color-70);
        }

        &.split-divider-default {
            background-color: var(--button-default-foreground-color);
        }

        &.split-divider-disruptive {
            background-color: var(--disruptive-color-70);
        }

        &.split-divider-neutral {
            background-color: var(--button-default-foreground-color);
        }
    }

    .split-divider-large {
        bottom: 10px;
        right: 50px;
        top: 10px;
    }

    .split-divider-medium {
        bottom: 8px;
        right: 42px;
        top: 8px;
    }

    .split-divider-small {
        bottom: 6px;
        right: 34px;
        top: 6px;
    }
}

.two-state-button {
    background-color: var(--button-two-state-default-background-color);
    color: var(--button-two-state-default-foreground-color);
    outline: var(--button-two-state-default-outline-color) solid 1px;
    outline-offset: -2px;

    .counter {
        background-color: var(--button-counter-default-background-color);
        display: inline-block;
        text-align: center;
        width: fit-content;
    }

    .button-text-1 {
        &.counter {
            margin: 0 $space-xs;
        }
    }

    .button-text-2 {
        &.counter {
            margin: 0 $space-xs;
        }
    }

    .button-text-3 {
        &.counter {
            margin: 0 $space-xxs;
        }
    }

    .two-state-button-content {
        columns: 2;
        flex-direction: column;

        .column {
            display: flex;
        }
    }

    &.button-stretch {
        .two-state-button-content {
            .column-one {
                width: calc(100% - 20px);
            }
        }
    }

    &:hover {
        background-color: var(--button-two-state-hover-background-color);
        color: var(--button-two-state-hover-foreground-color);
        outline-color: var(--button-two-state-hover-outline-color);

        .counter {
            background-color: var(--button-counter-hover-background-color);
        }
    }

    &:active {
        background-color: var(--button-two-state-active-background-color);
        color: var(--button-two-state-active-foreground-color);
        outline-color: var(--button-two-state-active-outline-color);

        .counter {
            background-color: var(--button-counter-active-background-color);
        }
    }

    &.checked {
        background-color: var(--button-two-state-checked-background-color);
        color: var(--button-two-state-checked-foreground-color);
        outline-color: var(--button-two-state-checked-outline-color);

        .counter {
            background-color: var(--button-counter-checked-background-color);
        }
    }

    &:focus {
        background-color: var(--button-two-state-focus-background-color);
        color: var(--button-two-state-focus-foreground-color);
        outline: 1px solid var(--button-two-state-focus-outline-color);

        &.checked {
            background-color: var(--button-two-state-checked-background-color);
            color: var(--button-two-state-checked-foreground-color);
            outline: 1px solid var(--button-two-state-focus-outline-color);
        }

        .counter {
            background-color: var(--button-counter-focus-background-color);
        }
    }
}

:global(.focus-visible) {
    .button {
        &:focus-visible {
            outline-width: $focus-visible-outline-width;
            outline-offset: $focus-visible-outline-offset-inner;
        }
    }

    .two-state-button {
        &:focus-visible {
            outline-width: $focus-visible-outline-width;
            outline-offset: $focus-visible-outline-offset-inner;

            &.checked {
                outline-width: $focus-visible-outline-width;
                outline-offset: $focus-visible-outline-offset-inner;
            }
        }
    }
}

.dark {
    .button-wrapper {
        .button {
            background-color: inherit;

            &.drop-shadow {
                box-shadow: 0 1px 2px rgba(15, 20, 31, 0.12),
                    0 2px 8px rgba(15, 20, 31, 0.16);
            }
        }
    }
}
