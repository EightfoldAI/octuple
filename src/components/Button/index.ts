export * from './Button.types';
export * from './BaseButton';
export * from './DefaultButton/DefaultButton';
export * from './NeutralButton/NeutralButton';
export * from './PrimaryButton/PrimaryButton';
export * from './SecondaryButton/SecondaryButton';
export * from './SplitButton/SplitButton';
export * from './TwoStateButton/TwoStateButton';
