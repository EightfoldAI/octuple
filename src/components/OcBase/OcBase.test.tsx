import React, { useMemo, useRef } from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Button } from '../Button';
import { render } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

const OcBaseComponent = (): JSX.Element => {
  const buttonRef: React.MutableRefObject<HTMLButtonElement> =
    useRef<HTMLButtonElement>(null);

  useMemo(() => {
    if (buttonRef.current) {
      buttonRef.current.name = 'ref-test';
    }
  }, []);

  return (
    <Button
      classNames="ocbase-class"
      data-test-id="test-ocbase-legacy"
      data-testid="test-ocbase"
      ref={buttonRef}
    />
  );
};

describe('OcBase', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Renders without crashing and uses data-testid', () => {
    const { container, getByTestId } = render(<OcBaseComponent />);
    const button = getByTestId('test-ocbase');
    expect(() => container).not.toThrowError();
    expect(button).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('Button accepts custom classes', () => {
    const { container } = render(<OcBaseComponent />);
    expect(container.querySelector('.ocbase-class')).toBeTruthy();
  });

  test('Button accepts legacy data-test-id', () => {
    const { container } = render(<OcBaseComponent />);
    expect(
      container
        .getElementsByClassName('ocbase-class')[0]
        .getAttribute('data-test-id')
    ).toBe('test-ocbase-legacy');
  });
});
