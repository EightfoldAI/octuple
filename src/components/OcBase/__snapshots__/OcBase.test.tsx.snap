// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OcBase Renders without crashing and uses data-testid 1`] = `
<div>
  <button
    aria-disabled="false"
    class="ocbase-class button button-default button-medium pill-shape"
    data-test-id="test-ocbase-legacy"
    data-testid="test-ocbase"
  >
    <span
      class="button-text-medium"
    >
      Button
    </span>
  </button>
</div>
`;
